bridge/tracker/tracker.go

```diff
@@ -0,0 +1,390 @@
+package tracker
+
+import (
+	"context"
+	"encoding/json"
+	"math/big"
+	"os"
+
+	"github.com/0xPolygon/polygon-edge/types"
+
+	"github.com/hashicorp/go-hclog"
+	"github.com/syndtr/goleveldb/leveldb"
+	"github.com/umbracle/go-web3"
+)
+
+const (
+	lastQueriedBlockNumber = "last-block-num" //	db key
+)
+
+//	Tracker represents an event listener that notifies
+//	for each event emitted on the rootchain (ethereum).
+//	Events of interest are defined in rootchain.go
+type Tracker struct {
+	logger hclog.Logger
+
+	// required block confirmations
+	confirmations uint64
+
+	// newHeads subscription channel
+	headerCh chan *ethHeader
+
+	// event channel
+	eventCh chan []byte
+
+	// cancel funcs
+	cancelSubscription, cancelEventTracking context.CancelFunc
+
+	//	rootchain client
+	client *rootchainClient
+
+	//	db for last processed block number
+	db *leveldb.DB
+}
+
+//	NewEventTracker returns a new tracker able to listen for events
+//	at the desired block depth. Events of interest are defined
+//	in rootchain.go
+func NewEventTracker(logger hclog.Logger, confirmations uint64) (*Tracker, error) {
+	tracker := &Tracker{
+		logger:        logger.Named("event_tracker"),
+		confirmations: confirmations,
+		headerCh:      make(chan *ethHeader, 1),
+		eventCh:       make(chan []byte),
+	}
+
+	//	create rootchain client
+	client, err := newRootchainClient(rootchainWS)
+	if err != nil {
+		logger.Error("cannot connect to rootchain", "err", err)
+
+		return nil, err
+	}
+
+	tracker.client = client
+
+	//	load db (last processed block number)
+	db, err := tracker.loadDB()
+	if err != nil {
+		logger.Error("cannot load db", "err", err)
+
+		return nil, err
+	}
+
+	tracker.db = db
+
+	return tracker, nil
+}
+
+//	Start runs the tracker's listening mechanism:
+//	1. startEventTracking - goroutine keeping track of
+//	the last block queried, as well as collecting events
+//	of interest.
+//
+//	2. startSubscription - goroutine responsible for handling
+//	the subscription object. This object is provided by the
+//	client's subscribeNewHeads method that initiates the subscription.
+func (t *Tracker) Start() error {
+	//	create cancellable contexts
+	ctxSubscription, cancelSubscription := context.WithCancel(context.Background())
+	t.cancelSubscription = cancelSubscription
+
+	ctxEventTracking, cancelEventTracking := context.WithCancel(context.Background())
+	t.cancelEventTracking = cancelEventTracking
```

Comment:
This can be a separate method

Reply:
I didn't want to further abstract the `Start()` method. That separate method would probably have to return 2 contexts and set the tracker's fields in its body - further hiding away the logic.

Reply:
You're directly modifying the `t.cancelSubscription` and `t.cancelEventTracking` fields, you can move away this block of code into a method called `initContexts` or something similar, that has no return values.

You keep track of the cancel funcs, but not of the contexts. Perhaps you could define a `cancelParams` type, and have 2 fields in `Tracker` called `ctxEventTracking` and `ctxSubscription` of type `cancelParams`, and you'd just use these throughout the code, instead of passing them around as params. It's weird that you keep the cancel method saved, but not the actual context it's tied to - by saving them both you make the code a lot cleaner

Reply:
[Refactored](https://github.com/0xPolygon/polygon-edge/pull/386/commits/22ed210d9e92dd02820edc89fd3ec5dc0e6cad92)